File: 프로그래머스/lv2/87946. 피로도/피로도.java
class Solution {
    static boolean[] visited;
    static int answer = 0;
    static int result=0;
    public int solution(int k, int[][] dungeons) {
        
        visited = new boolean[dungeons.length];

        dfs(k, 0, dungeons, 0);
        
        return answer;
    }
    
    static void dfs(int k, int idx, int[][] dungeons, int count){
        answer = Math.max(answer, count);
        for(int i=0; i<dungeons.length; i++){
            if(!visited[i] && k >= dungeons[i][0]){      // 최소 필요 피로도
                visited[i] = true;
                dfs(k-dungeons[i][1], i+1, dungeons, count+1);
                visited[i] = false;
            }
        }     
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    global answer 
    answer = -1
    
    global visited 
    visited = [False]*len(dungeons)
    
    answer = dfs(dungeons, 0, k, 0)
    
    return answer

def dfs(dungeons, num, k, count) -> int:
    global answer
    answer = max(answer, count)
    
    for i in range(len(dungeons)):
        if not visited[i] and dungeons[i][0] <= k:
            visited[i] = True
            dfs(dungeons, i+1, k-dungeons[i][1], count+1)
            visited[i] = False
    return answer